integration/commands_administration_test.go

```diff
@@ -773,6 +775,227 @@ func TestCommandsAdministrationDBStatsEmptyWithScale(t *testing.T) {
+			},
+		},
+		// this confirms that after we rename foo to bar and then create foo again,
+		// 1. bool-false doesn't exist
+		// 2. the newly inserted documents exist
+		"InsertIntoOld": {
+			collection:       "foo",
+			targetCollection: "bar",
+			expected:         bson.D{{"ok", float64(1)}},
+			toInsert: []interface{}{
+				bson.D{{"_id", 1}},
+				bson.D{{"_id", 2}},
+			},
+		},
+		"RenameMaxTableLen": {
+			collection:       "foo",
+			targetCollection: maxTableLen + "a",
+			err: &mongo.CommandError{
+				Code:    73,
+				Name:    "InvalidNamespace",
+				Message: "collection is too long",
+			},
+		},
+	} {
+		name, tc := name, tc
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+
+			ctx, collection := setup.Setup(t, shareddata.Bools)
+
+			db := collection.Database()
+
+			require.NotEmpty(t, insertCollections)
+			for _, coll := range insertCollections {
+				require.NoError(t, db.CreateCollection(ctx, coll))
+			}
+
+			var actual bson.D
+
+			sourceNamespace := fmt.Sprintf("%s.%s", db.Name(), tc.collection)
+			targetNamespace := fmt.Sprintf("%s.%s", db.Name(), tc.targetCollection)
+
+			cmd := bson.D{
+				{"renameCollection", sourceNamespace},
+				{"to", targetNamespace},
+			}
+
+			err := collection.Database().Client().Database(db.Name()).RunCommand(
+				ctx, cmd,
+			).Decode(&actual)
+
+			if tc.err != nil {
+				AssertEqualError(t, *tc.err, err)
+				return
+			}
+
+			require.NoError(t, err)
+			require.Equal(t, tc.expected, actual)
+
+			collections, err := db.ListCollectionNames(
+				ctx,
+				bson.D{},
+				nil,
+			)
+
+			require.NoError(t, err)
+
+			require.Contains(t, collections, tc.targetCollection)
+			require.NotContains(t, collections, tc.collection)
+
+			if tc.toInsert != nil {
+				require.Equal(t, "bar", tc.targetCollection)
+				require.NoError(t, db.CreateCollection(ctx, tc.collection))
+
+				res, err := db.Collection(tc.collection).InsertMany(ctx, tc.toInsert)
+				require.NoError(t, err)
+				require.Equal(
+					t,
+					&mongo.InsertManyResult{[]interface{}{int32(1), int32(2)}},
+					res,
+				)
+
+				var v any
+				err = db.Collection(tc.targetCollection).FindOne(
+					ctx, bson.D{{"bool-false", false}},
+				).Decode(&v)
+
+				require.ErrorIs(t, err, mongo.ErrNoDocuments)
+			}
+		})
+	}
+}
+
+func TestCommandsAdministrationRenameCollectionStress(t *testing.T) {
+	setup.SkipForTigrisWithReason(t, "Rename is not supported for Tigris")
+
+	ctx, collection := setup.Setup(t) // no providers there, we will create collections that are needed
+	db := collection.Database()
+
+	collNum := runtime.GOMAXPROCS(-1) * 10
+
+	// create collections that we will attempt to rename later
+	for i := 0; i < collNum; i++ {
+		collName := fmt.Sprintf("rename_collection_stress_%d", i)
+		err := db.CreateCollection(ctx, collName)
+		require.NoError(t, err)
+	}
+
+	ready := make(chan struct{}, collNum)
+	start := make(chan struct{})
+
+	var errNum atomic.Int32
+	renamedCollections := map[string]struct{}{}
```

Comment:
maybe `atomic.Value` is what you are looking for https://pkg.go.dev/sync/atomic#Value. No need to have mutex for that. :hugs:

Reply:
Will add a comment to explain why we chose this approach as mentioned on call! 

:)